Refactor extension to use an inhibitor manager

stuarthayhurst opened the pull request:
This is a fairly extensive rewrite of `extension.js`, and I'm not proud of putting it in 1 commit, but absolutely nothing would work in the slightest until everything was moved over to an inhibitor manager, as I suggested in #285.

The current solution seems to create an inhibitor for every excuse to inhibit, which has lead to the extension losing track of its inhibitors in the past, and suffered from race conditions too. The new solution abstracts the inhibiting away to a new class, `InhibitorManager`. It connects to signals and calls `_updateState()`, which uses `_getInhibitReasons()` to scan every excuse to inhibit (user toggle, apps, fullscreen). It then compares these against the user's settings, and decides whether to inhibit or not, and updates the inhibitor and night light accordingly. This means we only every have 1 inhibitor, so we can't easily lose track of it. It also makes the app handling simpler, as we just add methods to handle the chosen app trigger, and return whichever app is responsible for triggering Caffeine.

This means adding new ways to enable and disable Caffeine just means making `_getInhibitReasons()` aware of them, making issues like #217, #307 and #330 much easier to approach.

In summary:
 - Slims down the codebase
 - Slims down DBus interfaces and GSettings schemas
 - Self-contains a lot of the code
 - Decouple the inhibitor logic and indicator logic
 - Creates the inhibitor synchronously, to put an end to the race conditions
 - Fixes #332 (I assume the old code got stuck in a loop of events)
 - Fixes a few miscellaneous bugs around triggering on workspaces

Any feedback and testing is appreciated :)

Closes #332

stuarthayhurst commented:
Finished my testing, nearly everything works, except the night light control is completely unresponsive and the extension quickly re-enables itself if you disable it with a trigger active, but I still need to compare that to the old behaviour.

stuarthayhurst commented:
Fixed the night light, any issues with this are new to me

stuarthayhurst commented:
Also, this removes the `toggle-state` settings key to control its state, since it didn't really fit into the new structure. That does mean we lose CLI control, but I'm not sure people actually used this as there are other desktop apps that would be better suited to it, and it kind of seemed like a side-effect of the old design.

stuarthayhurst commented:
Apparently this closes #332 as well

stuarthayhurst commented:
@pakaoraki @eonpatapon GNOME 47.alpha is out, I'll start work on porting soon, but ideally this would be merged by then. Any chance of a quick review?

Thanks :)

pakaoraki commented:
Sorry for the late answer. I think it's a really nice improvement, I like this new approach. For now, this version is not working on my computer (Ubuntu 24/Fedora 40, Gnome 46): I have this error message "TypeError: app is null". I will test/debug it soon.

stuarthayhurst commented:
Do you have your app trigger set to active workspace only? I think I missed a check there.

pakaoraki commented:
I ran this version for the last few days.
I find some issues: 
- Apps trigger not working: it create multiple inhibitors again. I reckon, you removed some mechanism I have previously added. For some reasons, apps events are not clean and can be erratic, calling multiple time `_updateState()` function in a row (for few millisecond). The concept of having one inhibitor flag to manage is really great but I don't think it can handle all the race condition problem due to this events.

- There is a minor problem with the subtitle label when a timer have been selected: when caffeine is off, the duration is displayed as subtitle: for some reason here it's not displayed immediately, you need to switch workspace (or sometimes close and reopen the toggle menu) to refresh the label.

Also, if it's possible, it could be nice to keep the `toggle-state` settings key, but from what I saw, it's seams a bit tricky to add it.

pakaoraki commented on caffeine@patapon.info/extension.js:
```diff
@@ -127,12 +106,302 @@ const AppsTrigger = {
+            }
+        }
+
+        // Update inhibitor if required
+        if (this._isInhibited !== shouldInhibit) {
+            if (shouldInhibit) {
+                this._addInhibitor(reasons);
+            } else {
+                this._removeInhibitor();
+            }
+        }
+
```
As it have only one inhibitor to manage, I recommend to destroy the previous inhibitor every time before creating the new one.  

I suggest to maybe store the cookie in a key (if it's possible) to make sure to remove it after the extension is badly stopped (crash). This can also be used to keep the inhibitor running after Caffeine been disabled/enabled when for example the screen is locked (it happened when using blank screen option).

stuarthayhurst replied:
Mmm, I wondered about doing that, but I left it for now since it *shouldn't* be possible to incorrectly call them, and if they're being called at the wrong time we have a bug anyway.

I'll sort the current issues, then better protect the inhibitor for extra safety.

stuarthayhurst commented:
> * Apps trigger not working: it create multiple inhibitors again. I reckon, you removed some mechanism I have previously added. For some reasons, apps events are not clean and can be erratic, calling multiple time `_updateState()` function in a row (for few millisecond). The concept of having one inhibitor flag to manage is really great but I don't think it can handle all the race condition problem due to this events.

I ran into this during development but thought I fixed this, I'll take another look, thanks

> * There is a minor problem with the subtitle label when a timer have been selected: when caffeine is off, the duration is displayed as subtitle: for some reason here it's not displayed immediately, you need to switch workspace (or sometimes close and reopen the toggle menu) to refresh the label.

I've got a fix for that, thanks

> Also, if it's possible, it could be nice to keep the `toggle-state` settings key, but from what I saw, it's seams a bit tricky to add it.

It shouldn't be too hard to implement it back, I dropped it since it seemed like an artefact of the previous design and less of an intended feature, it allowed slimming down the gsettings schema and it was one less entry point to deal with. But if people miss it it can be added back, I'll tackle that after the new system works properly.

stuarthayhurst commented:
Since JavaScript is single-threaded, we can just test if we're waiting to create an inhibitor, as well as if we've already created one. If we need to delete and inhibitor that's waiting to be created, we can check for this in the callback to DBus, and just destroy it straight away.

This hasn't broken anything on my system, but I also don't know if it fixed the issue, since I couldn't reproduce it.

stuarthayhurst commented:
Hopefully that should be any actual bugs fixed. The trigger signals could be improved to connect based off of what it actually needs, and I'll look for an elegant way to bring back `toggle-state`.

stuarthayhurst commented on caffeine@patapon.info/extension.js:
```diff
@@ -127,12 +106,302 @@ const AppsTrigger = {
+        global.display.connectObject('notify::focus-window', () => this._updateState(), this);
+        global.workspace_manager.connectObject('workspace-switched',
+            () => this._updateState(), this);
+
```
Implemented relevant signals suggestion. I'm not convinced on the signal blocker, since if we have an event happen and then revert in a short space of time, the revert would be missed and it would be stuck in the old state. I think that's just going to cause bugs down the line.

Filtering the app id feels a little redundant, as it would be trying to do part of `_updateState()`'s job for it, I think it's better to leave it as a sort of sync method, that just takes in the environment around it when given a nudge.

I'm happy that this point is resolved, if you're not just let me know :)

stuarthayhurst commented:
Additionally, this fixes a bug that moving an app between workspaces would sometimes be missed depending on which app was focused. I don't think that had been reported yet.

stuarthayhurst commented:
@pakaoraki Rebased since it's been a while, what's your position on this? I haven't had any issues on my system, but then again you found some bugs before that I couldn't reproduce.

stuarthayhurst commented:
Found a way to call the `Inhibit` method synchronously, so now the extension only has to manage a single inhibitor, synchronously. That should put an end to the race conditions, and also drops all the ugly code to try and work around it.

Let me know if anyone still has any issues with anything :)

pakaoraki commented:
Sorry, I've been really busy since weeks now. I will try again the code and review it again as soon as I can (before the weekend). Please test it fully apps trigger mode with different scenarios (open multiple apps, change virtual desktop, etc...) last time, multiple flags were created.

stuarthayhurst commented:
> Sorry, I've been really busy since weeks now. I will try again the code and review it again as soon as I can (before the weekend).

Yeah no worries, I've only just found time to pick it back up anyway

> Please test it fully apps trigger mode with different scenarios (open multiple apps, change virtual desktop, etc...) last time, multiple flags were created.

I'm running it on my machine and I haven't found any issues so far, I'll try to be as comprehensive as possible. I can't really test the night light as easily though, since it crashes my machine (NVIDIA at it again...)

pakaoraki commented:
Tested on last stable Fedora and Ubuntu, it's working great, no more issues with multiple apps flags. I will keep testing today but everything seams fine so far, thanks for this amazing work ! 
I let @eonpatapon to review it maybe as it is a big rewrite.

stuarthayhurst commented:
This would be great to get merged before GNOME 48 is ready, then it would benefit from all the testing done on that version.

@eonpatapon Do you think you'll get time to review this? No worries if you can't, I know how busy everyone can be :)
